tool/teleport/common/sftp.go

```diff
@@ -75,17 +79,96 @@ func (c compositeCh) Close() error {
 	return trace.NewAggregate(c.r.Close(), c.w.Close())
 }
 
+type byteReader struct {
+	buf []byte
+	io.Reader
+}
+
+func newByteReader(r io.Reader) *byteReader {
+	return &byteReader{
+		Reader: r,
+		buf:    make([]byte, 1),
+	}
+}
+
+func (r *byteReader) ReadByte() (byte, error) {
+	_, err := r.Reader.Read(r.buf)
+	if err != nil {
+		return 0, err
+	}
+
+	return r.buf[0], nil
+}
+
+type allowedOps struct {
+	write bool
+	path  string
+}
+
 // sftpHandler provides handlers for a SFTP server.
 type sftpHandler struct {
-	logger *log.Entry
-	events chan<- *apievents.SFTP
+	logger  *log.Entry
+	allowed *allowedOps
```

Comment:
Do we need a pointer here? `allowedOps` looks pretty small.

Reply:
We don't *need* it, the main reason was to make it easier to check for the zero value or not. I don't think it'd have a major perf or memory impact either way though.